src/moonlink/src/storage/mooncake_table.rs

```diff
@@ -754,84 +763,172 @@ impl MooncakeTable {
         self.mem_slice.get_num_rows() >= self.metadata.config.mem_slice_size
     }
 
-    /// Flush `mem_slice` into parquet files and return the resulting `DiskSliceWriter`.
-    ///
-    /// When `snapshot_task` is provided, new batches and indices are recorded so
-    /// that they can be included in the next snapshot.  The `lsn` parameter
-    /// specifies the commit LSN for the flushed data.  When `lsn` is `None` the
-    /// caller is responsible for setting the final LSN on the returned
-    /// `DiskSliceWriter`.
-    async fn flush_mem_slice(
-        mem_slice: &mut MemSlice,
-        metadata: &Arc<TableMetadata>,
-        next_file_id: u32,
-        lsn: Option<u64>,
-        snapshot_task: Option<&mut SnapshotTask>,
-    ) -> Result<DiskSliceWriter> {
+    /// Drains the current mem slice and prepares a disk slice for flushing.
+    /// Adds current mem slice batches and indices to `next_snapshot_task`.
+    fn prepare_disk_slice(&mut self, lsn: u64) -> Result<DiskSliceWriter> {
         // Finalize the current batch (if needed)
-        let (new_batch, batches, index) = mem_slice.drain().unwrap();
+        let (new_batch, batches, index) = self.mem_slice.drain()?;
 
         let index = Arc::new(index);
-        if let Some(task) = snapshot_task {
-            if let Some(batch) = new_batch {
-                task.new_record_batches.push((batch.0, batch.1, None));
-            }
-            task.new_mem_indices.push(index.clone());
+        if let Some(batch) = new_batch {
+            self.next_snapshot_task
+                .new_record_batches
+                .push((batch.0, batch.1, None));
```

Comment:
let me try again, do you mind we using struct instead of tuples? Struct provides field naming, which is much more readable.